feat(tencentcloud pipeline): add disableCluster, disableServerGroup, …

zexiplus opened the pull request:
This PR aims to add 7 pipeline stages includes disableCluster, disableServerGroup, enableServerGroup, rollbackCluster, scaleDownCluster, shrinkCluster, tagImage for Tencent cloud. Tencent cloud provider proposal is spinnaker/governance#79

The preview is as follows：
![disableCluster](https://user-images.githubusercontent.com/15920861/88024907-135a2900-cb66-11ea-937c-d45170cffaaf.png)
![disableServerGroup](https://user-images.githubusercontent.com/15920861/88024987-3258bb00-cb66-11ea-9cee-ffbd7758b673.png)
![enableServerGroup](https://user-images.githubusercontent.com/15920861/88025045-48ff1200-cb66-11ea-8675-28da5fadb468.png)
![rollbackCluster](https://user-images.githubusercontent.com/15920861/88025102-5d430f00-cb66-11ea-94af-13b34a6f8d62.png)
![scaleDownCluster](https://user-images.githubusercontent.com/15920861/88025150-6c29c180-cb66-11ea-845f-7d5dae98ecf5.png)
![shrinkCluster](https://user-images.githubusercontent.com/15920861/88025185-7cda3780-cb66-11ea-94cc-1df0ce372842.png)
![tagImage](https://user-images.githubusercontent.com/15920861/88025214-8b285380-cb66-11ea-9f65-2f5b43545df7.png)

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/DisableCluster.tsx:
```diff
@@ -0,0 +1,122 @@
+import * as React from 'react';
```
Please organize imports according to Deck's [import order conventions](https://github.com/spinnaker/deck/wiki/Conventions#imports). (Same comment for all other files.)

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/DisableCluster.tsx:
```diff
@@ -0,0 +1,122 @@
+
+interface IProps extends IStageConfigProps {}
+interface IState {
+  accounts: any[];
```
Can we avoid using `any` here in favor of a strongly-typed tencentcloud account? (Same comment for all other any-typed accounts in other files.)

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/DisableCluster.tsx:
```diff
@@ -0,0 +1,122 @@
+
+    if (
+      stage.isNew &&
+      get(application, 'attributes.platformHealthOnlyShowOverride') &&
```
Can likely remove some of these lodash `get`s in favor of Typescript [optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining). (Same comment for other uses of `get` in other files.)

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/disableClusterStage.js:
```diff
@@ -0,0 +1,24 @@
+'use strict';
```
All new files must be added in Typescript.

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/disableClusterStage.js:
```diff
@@ -0,0 +1,24 @@
+import { DisableCluster } from './DisableCluster';
+
+export const TENCENTCLOUD_PIPELINE_STAGES_DISABLE_CLUSTER = 'spinnaker.tencentcloud.pipeline.stage.disableClusterStage';
+angular.module(TENCENTCLOUD_PIPELINE_STAGES_DISABLE_CLUSTER, []).config(function() {
```
You shouldn't need to register any of these stages as Angular modules, as long as each file that registers a stage with the pipeline registry is imported to the tencentcloud entrypoint (`tencentcloud.module.ts`).

maggieneterval commented on app/scripts/modules/tencentcloud/src/tencentcloud.module.ts:
```diff
@@ -9,14 +9,33 @@ import { TENCENTCLOUD_SEARCH_SEARCHRESULTFORMATTER } from './search/searchResult
 import './validation/ApplicationNameValidator';
 
+// load pipeline stage
+import { TENCENTCLOUD_PIPELINE_STAGES_ENABLE_SG } from './pipeline/stages/enableServerGroup/enableSgStage';
```
```suggestion
import './pipeline/stages/enableServerGroup/enableSgStage';
```
Alternately, can export all stages from a single `./pipeline/stages/index` and just import that here. (I think this is what the kubernetes provider does if you want to look at an example.)

maggieneterval commented on app/scripts/modules/tencentcloud/src/tencentcloud.module.ts:
```diff
@@ -9,14 +9,33 @@ import { TENCENTCLOUD_SEARCH_SEARCHRESULTFORMATTER } from './search/searchResult
 
 export const TENCENTCLOUD_MODULE = 'spinnaker.tencentcloud';
-module(TENCENTCLOUD_MODULE, [TENCENTCLOUD_REACT_MODULE, TENCENTCLOUD_SEARCH_SEARCHRESULTFORMATTER]).config(() => {
+module(TENCENTCLOUD_MODULE, [
```
Can remove (see other comments).

zexiplus commented:
@maggieneterval Thanks for your review, I will consider refactoring this PR

zexiplus commented:
@maggieneterval Hi, I have already refactored this PR(I added four stages includes disable cluster, rollback cluster, scaledown cluster, shrink cluster). This time I removed some angular module definition, use functional components,  and so on. Looking forward to your more suggestions

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/disableClusterStage.ts:
```diff
@@ -0,0 +1,11 @@
+  cloudProvider: 'tencentcloud',
+  component: DisableClusterConfig,
+  validateFn: validate,
+} as IStageTypeConfig);
```
Why do we need this `as` assertion? Are we missing any expected properties? (Same question for other stage configs in this PR).

zexiplus replied:
I added it by mistake, now I have removed them at all

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/DisableClusterStageForm.tsx:
```diff
@@ -0,0 +1,105 @@
+        <div className="form-inline">
+          Keep the
+          <FormikFormField
+            name={'remainingEnabledServerGroups'}
```
```suggestion
            name="remainingEnabledServerGroups"
```
I think this can be simplified to just a string literal here & a few other places in this PR

zexiplus replied:
good suggestion

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/disableCluster/DisableClusterStageForm.tsx:
```diff
@@ -0,0 +1,105 @@
+          <FormikFormField
+            name={'remainingEnabledServerGroups'}
+            required
+            input={() => (
```
There are several components such as [`SelectInput`](https://github.com/spinnaker/deck/blob/master/app/scripts/modules/core/src/presentation/forms/inputs/SelectInput.tsx) that can be found in deck's `/core/presentation` module that are intended for use with `FormikFormField` and handle updating the key specified as the field's `name` (in this case, `remainingEnabledServerGroups`) so that you don't need to use a vanilla `input` and write your own `value` or `onChange` handler. (Same comment for other `inputs` in this PR.)

zexiplus replied:
Fixed it

zexiplus commented:
> Thank you so much for taking the time to update this PR to use Deck's newer stage config patterns; looks great! I left just a couple more comments inline; please let me know if you need any help resolving them.
> 
> I also wanted to highlight that once you have submitted 5 significant PRs, you are eligible to apply to become a Spinnaker [Reviewer](https://github.com/spinnaker/governance/issues/new?assignees=ajordens%2C+ethanfrogers%2C+ezimanyi%2C+plumpy%2C+robzienert&labels=committee%2Ftechnical-oversight&template=reviewer-application.md&title=REQUEST%3A+New+Reviewer+status+for+%3Cyour-GH-username%3E). It would be great to get to a point where you or one of the other Tencent SIG leads (cc @shengyu @donghun221) were Reviewers and could take a first pass at all Tencent-specific contributions, since @christopherthielen and I have limited context on that cloud provider. That said, more than happy to help with these first few PRs and sponsor you on your Reviewer application to vouch for your contributions when you are ready to apply.

Thanks for your suggestion again, I changed the code as you suggested. I am willing to apply to be a reviewer after this PR merged.

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/shrinkCluster/shrinkClusterStage.ts:
```diff
@@ -0,0 +1,11 @@
+  cloudProvider: 'tencentcloud',
+  component: ShrinkClusterConfig,
+  validateFn: validate,
+} as IStageTypeConfig);
```
```suggestion
});
```
I think we just missed one of these `as IStageTypeConfig` to remove.

maggieneterval commented:
> Thanks for your suggestion again, I changed the code as you suggested. I am willing to apply to be a reviewer after this PR merged.

Great, this looks ready to merge after we remove the final unnecessary `as` assertion.

In the future, it's a bit easier for reviewers to take a second pass at PRs when you address code review comments in a separate commit rather than squashing them into the original commit. Thanks!

zexiplus commented:
> > Thanks for your suggestion again, I changed the code as you suggested. I am willing to apply to be a reviewer after this PR merged.
> 
> Great, this looks ready to merge after we remove the final unnecessary `as` assertion.
> 
> In the future, it's a bit easier for reviewers to take a second pass at PRs when you address code review comments in a separate commit rather than squashing them into the original commit. Thanks!

got it

maggieneterval commented on app/scripts/modules/tencentcloud/src/pipeline/stages/shrinkCluster/shrinkClusterStage.ts:
```diff
@@ -0,0 +1,11 @@
+import { validate } from './shrinkClusterValidators';
+import { Registry, IStageTypeConfig } from '@spinnaker/core';
```
```suggestion
import { Registry } from '@spinnaker/core';
```
I think now CI is failing because of this unused import.

zexiplus replied:
yup, removed
